ledger/src/snapshot_utils.rs

```diff
@@ -592,8 +599,13 @@ where
         &root_paths.snapshot_file_path,
         MAX_SNAPSHOT_DATA_FILE_SIZE,
         |stream| {
-            let mut bank: Bank = match snapshot_version {
-                SNAPSHOT_VERSION => deserialize_from_snapshot(stream.by_ref())?,
+            let versioned_bank: BankVersions = match snapshot_version {
+                SNAPSHOT_VERSION_1_0 => {
+                    BankVersions::Bank1_0(deserialize_from_snapshot(stream.by_ref())?)
+                }
+                SNAPSHOT_VERSION_1_1 => {
+                    BankVersions::Bank1_1(deserialize_from_snapshot(stream.by_ref())?)
```

Comment:
@mvines, @sakridge, I ran ledger-tool `verify` on an old 1.0 snapshot and this works fine.

At first I was like great, the change works!

But on second thought, isn't this a vulnerability? I fundamentally changed a field in the bank that's used by many parts of the system, and the snapshot hash still verified fine...

Reply:
We don't require the Bank struct be immutable.  There could exist another Solana implementation written in bash that produces the same snapshot hash as the current mostly-Rust implementation

Reply:
Seems like critical fields in the bank that should be verified(please check if any are missing!)

1) `parent_hash`
2) `hard_forks`
3) `transaction_count`
4) `tick_height`
5) `signature_count`
6) `capitalization`
7) `max_tick_height`
8) `hashes_per_tick`
9) `ticks_per_slot`
10) `ns_per_slot`
11) `genesis_creation_time`
12) `slots_per_year`
13) `slots_per_segment`
14) `slot`
15) `epoch`
16) `block_height`
17) `collector_id`
18) `collector_fees`
19) `fee_calculator`
20) `fee_rate_governor`
21) `collected_rent`
22) `rent_collector`
23) `epoch_schedule`
24) `inflation`
25) `stakes`
26) `storage_accounts`
27) `epoch_stakes`
28) `is_delta`
29) `message_processor`

Reply:
@mvines seems like I could just replace a stake account, or a epoch stakes account, or inflation/rent and it would affect consensus

Reply:
That would cause a bank hash mismatch

Reply:
That's part of https://github.com/solana-labs/solana/issues/7167 and on @ryoqun 's radar. Part of why we are using the flags to get verified snapshots from only trusted nodes.

Reply:
Ultimately all those changes would be reflected in the contents of some account(s), which are hashed

Reply:
@mvines, does the bank hash take into account fields that are cached in the bank? I only see a few things like parent hash, accounts delta, signature count: https://github.com/solana-labs/solana/blob/master/runtime/src/bank.rs#L1851-L1855

Reply:
@sakridge ah ok, thanks!

Reply:
@mvines not necessarily, as some cached state like the cached `EpochStakes` in the bank could be changed to have drastically different balances, while the account state itself is correct. Then people wouldn't notice they were using the wrong stakes until they update to a new epoch in the `EpochStakes` 2 epochs later

Reply:
Hashing all bank fields will be unnecessary and slow though, we don't need that

Reply:
> Then people wouldn't notice they were using the wrong stakes until they update to a new epoch in the EpochStakes 2 epochs later

At which point consensus will correctly fail since accounts will be mismatched.   That's not a correctness issue

Reply:
@mvines, so for example, let's say we were using `EpochStakes` as the source of "weighting" for various consensus mechanisms. 

Among other things like gossip, turbine, etc. the most drastic example of this would be say, we change to, as @aeyakovenko has suggested in the past, using `EpochStakes` balances for calculating bank weight.

```
A: 1 
B: 1
C: 1
```

and `A` sends a new validator `D` a snapshot with `EpochStakes`:

```
A: 100
B: 0
C: 0
```

Then `D` will gladly lock itself out very fast on whatever fork `A` generates as it thinks the majority of the network is with `A`

Reply:
Not sure that hashing all the things is the answer though, we need to leave some freedom for the implementation to roam.  Maybe hashing one or two more things is the answer though

Reply:
yeah agreed! Ideally some subset of that giant list that's being used in "critical" areas. Maintaining a subset scares me though because any one of those fields could suddenly become "critical" when new logic is pushed :P. 

Even things you may want to vary based on implementation like `hashes_per_tick` or `ticks_per_slot` scare me, as you can cause people to generate bad slots/lose their leader rewards/stall. If we hash everything we don't miss anything.

Reply:
>  If we hash everything we don't miss anything.
And can then never change anything :)

Reply:
> 
> 
> > If we hash everything we don't miss anything.
> 
> And can then never change anything :)

Why can't you just recompute the hash value with the changes?

Reply:
Speed is not a reason to do it, the whole serialized size of the bank is 200KB, that should take about 500us to hash with sha-2. We are already spending 100s of ms to hash the account states.

Reply:
We can, and we'll probably need one of these painful snapshot/Bank adaption PRs every time we do so.

Reply:
That, or you re-compute these values from the account state which is hashed. That seems like more work to me.

Reply:
hehe I'm just saying that I'm pretty sure that all of the 29 Bank fields that Carl listed above do not all need be included in the bank hash.

Reply:
> That's part of #7167 and on @ryoqun 's radar. Part of why we are using the flags to get verified snapshots from only trusted nodes.

@carllin FYI: As @sakridge said, this is the wip-stale-closed PR for the exact thing of this discussion: #8185. I'm steadily approaching to it. There are still other bunch of security issues in the snapshot to tackle first, though. ;)